File: backend/src/domain/services/MatchingService.ts
import { HelpRequest } from '../entities/HelpRequest';
import { HelpOffer } from '../entities/HelpOffer';

export interface MatchResult {
  request: HelpRequest;
  offer: HelpOffer;
  distanceKm: number;
  score: number; // higher = better match
}

/**
 * Domain service for matching help requests with help offers.
 * Pure domain logic — no infrastructure dependencies.
 */
export class MatchingService {
  /**
   * Find the best available offers for a given request.
   * Scores based on: type match, distance, capacity, priority.
   */
  public findBestMatches(
    request: HelpRequest,
    offers: HelpOffer[],
    maxResults: number = 5
  ): MatchResult[] {
    const availableOffers = offers.filter(
      (o) => o.isAvailable() && o.type === request.type
    );

    const scored = availableOffers.map((offer) => {
      const distance = this.calculateDistance(
        request.latitude,
        request.longitude,
        offer.latitude,
        offer.longitude
      );

      const score = this.calculateScore(request, offer, distance);

      return { request, offer, distanceKm: distance, score };
    });

    return scored
      .sort((a, b) => b.score - a.score)
      .slice(0, maxResults);
  }

  /**
   * Auto-match all open requests with available offers for an emergency.
   */
  public autoMatch(
    requests: HelpRequest[],
    offers: HelpOffer[]
  ): MatchResult[] {
    const matches: MatchResult[] = [];
    const usedOfferIds = new Set<string>();
    const usedRequestIds = new Set<string>();

    // Sort requests by priority (urgent first)
    const sortedRequests = [...requests]
      .filter((r) => r.isOpen())
      .sort((a, b) => {
        const priorityOrder = { URGENT: 0, HIGH: 1, MEDIUM: 2, LOW: 3 };
        return priorityOrder[a.priority] - priorityOrder[b.priority];
      });

    for (const request of sortedRequests) {
      if (usedRequestIds.has(request.id)) continue;

      const availableOffers = offers.filter(
        (o) => o.isAvailable() && !usedOfferIds.has(o.id)
      );

      const bestMatches = this.findBestMatches(request, availableOffers, 1);

      if (bestMatches.length > 0) {
        const best = bestMatches[0];
        matches.push(best);
        usedOfferIds.add(best.offer.id);
        usedRequestIds.add(request.id);
      }
    }

    return matches;
  }

  private calculateScore(
    request: HelpRequest,
    offer: HelpOffer,
    distanceKm: number
  ): number {
    let score = 100;

    // Distance penalty (closer is better)
    score -= distanceKm * 2;

    // Capacity bonus
    if (offer.capacity >= request.numberOfPeople) {
      score += 20;
    }

    // Priority bonus (urgent requests get higher scoring matches)
    const priorityBonus: Record<string, number> = {
      URGENT: 30,
      HIGH: 20,
      MEDIUM: 10,
      LOW: 0,
    };
    score += priorityBonus[request.priority] ?? 0;

    return Math.max(0, score);
  }

  private calculateDistance(
    lat1: number,
    lng1: number,
    lat2: number,
    lng2: number
  ): number {
    const R = 6371;
    const dLat = ((lat2 - lat1) * Math.PI) / 180;
    const dLng = ((lng2 - lng1) * Math.PI) / 180;
    const a =
      Math.sin(dLat / 2) * Math.sin(dLat / 2) +
      Math.cos((lat1 * Math.PI) / 180) *
        Math.cos((lat2 * Math.PI) / 180) *
        Math.sin(dLng / 2) *
        Math.sin(dLng / 2);
    const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
    return R * c;
  }
}

File: backend-dotnet/src/Hestia.Domain/Services/MatchingService.cs
using Hestia.Domain.Entities;
using Hestia.Domain.Enums;

namespace Hestia.Domain.Services;

/// <summary>
/// Domain service for matching help requests with help offers.
/// Pure domain logic — no infrastructure dependencies.
/// </summary>
public class MatchingService
{
    /// <summary>
    /// Find the best available offers for a given request.
    /// Scores based on: type match, distance, capacity, priority.
    /// </summary>
    public IReadOnlyList<MatchResult> FindBestMatches(
        HelpRequest request,
        IEnumerable<HelpOffer> offers,
        int maxResults = 5)
    {
        var availableOffers = offers
            .Where(o => o.IsAvailable() && o.Type == request.Type)
            .ToList();

        var scored = availableOffers.Select(offer =>
        {
            var distance = CalculateDistance(
                request.Latitude, request.Longitude,
                offer.Latitude, offer.Longitude);
            var score = CalculateScore(request, offer, distance);
            return new MatchResult(request, offer, distance, score);
        });

        return scored
            .OrderByDescending(r => r.Score)
            .Take(maxResults)
            .ToList();
    }

    /// <summary>
    /// Auto-match all open requests with available offers for an emergency.
    /// Prioritises urgent requests first.
    /// </summary>
    public IReadOnlyList<MatchResult> AutoMatch(
        IEnumerable<HelpRequest> requests,
        IEnumerable<HelpOffer> offers)
    {
        var results = new List<MatchResult>();
        var usedOfferIds = new HashSet<Guid>();
        var usedRequestIds = new HashSet<Guid>();
        var offersList = offers.ToList();

        var priorityOrder = new Dictionary<HelpRequestPriority, int>
        {
            [HelpRequestPriority.Urgent] = 0,
            [HelpRequestPriority.High] = 1,
            [HelpRequestPriority.Medium] = 2,
            [HelpRequestPriority.Low] = 3,
        };

        var sortedRequests = requests
            .Where(r => r.IsOpen())
            .OrderBy(r => priorityOrder[r.Priority])
            .ToList();

        foreach (var request in sortedRequests)
        {
            if (usedRequestIds.Contains(request.Id)) continue;

            var available = offersList.Where(o => o.IsAvailable() && !usedOfferIds.Contains(o.Id));
            var best = FindBestMatches(request, available, 1);

            if (best.Count > 0)
            {
                results.Add(best[0]);
                usedOfferIds.Add(best[0].Offer.Id);
                usedRequestIds.Add(request.Id);
            }
        }

        return results;
    }

    private static double CalculateScore(HelpRequest request, HelpOffer offer, double distanceKm)
    {
        var score = 100.0;

        // Distance penalty (closer is better)
        score -= distanceKm * 2;

        // Capacity bonus
        if (offer.Capacity >= request.NumberOfPeople)
            score += 20;

        // Priority bonus
        score += request.Priority switch
        {
            HelpRequestPriority.Urgent => 30,
            HelpRequestPriority.High => 20,
            HelpRequestPriority.Medium => 10,
            _ => 0
        };

        return Math.Max(0, score);
    }

    private static double CalculateDistance(double lat1, double lng1, double lat2, double lng2)
    {
        const double R = 6371;
        var dLat = (lat2 - lat1) * Math.PI / 180;
        var dLng = (lng2 - lng1) * Math.PI / 180;
        var a = Math.Sin(dLat / 2) * Math.Sin(dLat / 2)
              + Math.Cos(lat1 * Math.PI / 180)
              * Math.Cos(lat2 * Math.PI / 180)
              * Math.Sin(dLng / 2) * Math.Sin(dLng / 2);
        var c = 2 * Math.Atan2(Math.Sqrt(a), Math.Sqrt(1 - a));
        return R * c;
    }
}

/// <summary>Represents a scored match between a request and an offer.</summary>
public record MatchResult(HelpRequest Request, HelpOffer Offer, double DistanceKm, double Score);
